port/posix/posix_transport_shm.c

```diff
@@ -1,301 +1,581 @@
+    do {
+        ret = posixTransportShm_UseMap( config->name, map);
+        if (    (ret == WH_ERROR_NOTFOUND) ||
+                (ret == WH_ERROR_NOTREADY) ) {
+            sleep(1);
+        } else {
+            break;
+        }
+    } while (retries-- > 0);
+
+    if (ret == WH_ERROR_OK) {
+        memset(ctx, 0, sizeof(*ctx));
+        strncpy(ctx->name,  config->name, sizeof(ctx->name));
+
+        /* Configure the underlying transport context */
+        tMemCfg->req_size  = map->header->req_size;
+        tMemCfg->resp_size = map->header->resp_size;
+        tMemCfg->req       = map->req;
+        tMemCfg->resp      = map->resp;
+
+        /* Initialize the shared memory transport */
+        ret = wh_TransportMem_InitClear(ctx->transportMemCtx, tMemCfg, connectcb,
+                                  connectcb_arg);
+
+        if (ret == WH_ERROR_OK) {
+            ctx->ptr = map->ptr;
+            ctx->size = map->size;
+            ctx->dma = map->dma;
+            ctx->dma_size = map->dma_size;
+        } else {
+            /* Problem initializing the transport */
+            (void)munmap(map->ptr, map->size);
+        }
+    }
+    return ret;
+#endif
 }
 
 
 int posixTransportShm_Cleanup(void* c)
 {
     posixTransportShmContext* ctx = (posixTransportShmContext*)c;
-    ShmHeader*                header;
 
-    if (ctx == NULL || ctx->transportMemCtx == NULL || ctx->shmBuf == NULL) {
+    if (ctx == NULL) {
         return WH_ERROR_BADARGS;
     }
 
-    header = (ShmHeader*)ctx->shmBuf;
-
-    /* Decrement the reference count */
-    if (atomic_fetch_sub(&header->refCount, 1) == 1) {
-        /* If the reference count is now zero, unlink the shared memory */
-        if (shm_unlink(ctx->shmObjName) == -1) {
-            perror("shm_unlink");
-            exit(EXIT_FAILURE);
-        }
-    }
+    if (ctx->ptr != NULL) {
+        ptshmHeader* header = (ptshmHeader*)ctx->ptr;
+        header->initialized = PTSHM_INITIALIZED_CLEANUP;
 
-    /* Unmap the shared memory */
-    if (munmap(ctx->shmBuf, sizeof(header) + ctx->transportMemCtx->req_size +
-                                ctx->transportMemCtx->resp_size) == -1) {
-        perror("munmap");
-        exit(EXIT_FAILURE);
+        (void)wh_TransportMem_Cleanup(ctx->transportMemCtx);
+        (void)munmap(ctx->ptr, ctx->size);
+        ctx->ptr = NULL;
     }
-
-    /* Free the allocated memory */
-    free(ctx->shmObjName);
-    free(ctx->transportMemCtx);
-
     return 0;
 }
 
 
 int posixTransportShm_SendRequest(void* c, uint16_t len, const void* data)
 {
     posixTransportShmContext* ctx = (posixTransportShmContext*)c;
+    int ret = WH_ERROR_OK;
 
     /* Only need to check NULL, mem transport checks other state info */
     if (ctx == NULL) {
         return WH_ERROR_BADARGS;
     }
 
-    return wh_TransportMem_SendRequest(ctx->transportMemCtx, len, data);
+    /* Check connected status */
+    switch(ctx->state) {
+    case PTSHM_STATE_NONE:
```

Comment:
A bit confused about this logic (as well as the client initialization. Since the client init is what initially zeros the context and calls `posixTransportShm_{Use/Handle}Map()` on it, and we don't ever want a user to be sending requests without first initializing the client context, should we even be handling the `NONE` state here? I'd have figured that this logic should have been in the `INITIALIZED` state, since that is what the client init handler sets the state to once it successfully initializes. Is this some sort of deferred initialization that is allowed if the original client init fails? If so perhaps we just call client init again here?

Reply:
The server creates and zeroes the object.  The client only has to map it.  NONE is used when the client was not able to open the object OR the object was not completely setup by the server.  This forces the client to attempt to reopen, map the header, and then remap the entire object. Once the client is able to successfully map the entire segment, it sets the intiitalzed member in the header and leaves the mapping intact.  Recall, the client init only tries once to open and map the object, so the entire sequence is tried again here as well.  Sorry for the misleading comments.